File: smartmoneyserver/src/main/java/fpt/aptech/server/dto/planned/PlannedTransactionRequest.java
package fpt.aptech.server.dto.planned;

import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotNull;
import lombok.Builder;

import java.math.BigDecimal;
import java.time.LocalDate;

@Builder
public record PlannedTransactionRequest(

        @NotNull(message = "Please select a wallet")
        Integer walletId,

        @NotNull(message = "Amount cannot be empty")
        @DecimalMin(value = "0.0", inclusive = false, message = "Amount must be greater than 0")
        BigDecimal amount,

        @NotNull(message = "Please select a category")
        Integer categoryId,

        // NULL nếu ctg không phải nợ/vay
        Integer debtId,

        String note,

        // 1: Bill | 2: Recurring
        @NotNull(message = "Invalid plan type")
        Integer planType,

        // 1: Ngày | 2: Tuần | 3: Tháng | 4: Năm
        @NotNull(message = "Please select repeat schedule")
        Integer repeatType,

        @NotNull(message = "Repeat interval cannot be empty")
        Integer repeatInterval,

        // Bitmask cho lặp tuần (CN=1,T2=2,T3=4,T4=8,T5=16,T6=32,T7=64)
        Integer repeatOnDayVal,

        @NotNull(message = "Please select start date")
        LocalDate beginDate,

        // "FOREVER" | "UNTIL_DATE" | "COUNT"
        @NotNull(message = "Please select end date option")
        String endDateOption,

        // Dùng khi endDateOption = "UNTIL_DATE"
        LocalDate endDateValue,

        // Dùng khi endDateOption = "COUNT" (số lần lặp)
        Integer repeatCount
) {}

File: APP MOBILE/smart_money/lib/modules/planned/models/planned_transaction_request.dart
/// Request tạo/sửa giao dịch lặp lại gửi lên server.
/// Tương ứng: PlannedTransactionRequest.java (server)
class PlannedTransactionRequest {
  final int walletId;
  final double amount;
  final int categoryId;
  final int? debtId;
  final String? note;
  final int planType;
  final int repeatType;
  final int repeatInterval;
  final int? repeatOnDayVal;
  final DateTime beginDate;
  final String endDateOption;
  final DateTime? endDateValue;
  final int? repeatCount;

  const PlannedTransactionRequest({
    required this.walletId,
    required this.amount,
    required this.categoryId,
    this.debtId,
    this.note,
    required this.planType,
    required this.repeatType,
    required this.repeatInterval,
    this.repeatOnDayVal,
    required this.beginDate,
    required this.endDateOption,
    this.endDateValue,
    this.repeatCount,
  });

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      'walletId': walletId,
      'amount': amount,
      'categoryId': categoryId,
      'planType': planType,
      'repeatType': repeatType,
      'repeatInterval': repeatInterval,
      'beginDate': _formatDate(beginDate),
      'endDateOption': endDateOption,
    };
    if (debtId != null) map['debtId'] = debtId;
    if (note != null) map['note'] = note;
    if (repeatOnDayVal != null) map['repeatOnDayVal'] = repeatOnDayVal;
    if (endDateValue != null) map['endDateValue'] = _formatDate(endDateValue!);
    if (repeatCount != null) map['repeatCount'] = repeatCount;
    return map;
  }

  static String _formatDate(DateTime dt) =>
      '${dt.year}-${dt.month.toString().padLeft(2, '0')}-${dt.day.toString().padLeft(2, '0')}';
}

